File: scripts/generate_teams_json.mjs
/**
 * Regenerates web/public/data/teams.json from data/MTeams.csv and data/WTeams.csv.
 * Run from repo root: node scripts/generate_teams_json.mjs
 * Or from web/: npm run generate:teams
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(__dirname, "..");

function readTeams(csvPath) {
  const text = fs.readFileSync(csvPath, "utf8");
  const lines = text.trim().split(/\r?\n/);
  const header = lines[0].split(",");
  const idIdx = header.indexOf("TeamID");
  const nameIdx = header.indexOf("TeamName");
  if (idIdx < 0 || nameIdx < 0) {
    throw new Error(`Missing TeamID/TeamName in ${csvPath}`);
  }
  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    const parts = lines[i].split(",");
    const id = Number.parseInt(parts[idIdx], 10);
    const name = String(parts[nameIdx] ?? "").trim();
    if (Number.isFinite(id)) {
      rows.push({ id, name });
    }
  }
  rows.sort((a, b) => a.name.localeCompare(b.name) || a.id - b.id);
  return rows;
}

const mPath = path.join(root, "data", "MTeams.csv");
const wPath = path.join(root, "data", "WTeams.csv");
if (!fs.existsSync(mPath) || !fs.existsSync(wPath)) {
  console.error(`Need ${mPath} and ${wPath}`);
  process.exit(1);
}

const payload = {
  generatedAt: new Date().toISOString(),
  sources: { M: "MTeams.csv", W: "WTeams.csv" },
  M: readTeams(mPath),
  W: readTeams(wPath),
};

const out = path.join(root, "web", "public", "data", "teams.json");
fs.mkdirSync(path.dirname(out), { recursive: true });
fs.writeFileSync(out, JSON.stringify(payload, null, 2) + "\n", "utf8");
console.log(
  `Wrote ${out} (${payload.M.length} men, ${payload.W.length} women teams)`,
);

File: scripts/generate_teams_json.py
"""
Emit web/public/data/teams.json from Kaggle MTeams.csv and WTeams.csv.

Run from repo root: python scripts/generate_teams_json.py
Requires: data/MTeams.csv, data/WTeams.csv
"""

from __future__ import annotations

import csv
import json
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DATA = ROOT / "data"
OUT = ROOT / "web" / "public" / "data" / "teams.json"


def _read_teams(path: Path) -> list[dict[str, int | str]]:
    rows: list[dict[str, int | str]] = []
    with path.open(newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for r in reader:
            tid = int(r["TeamID"])
            name = str(r["TeamName"]).strip()
            rows.append({"id": tid, "name": name})
    rows.sort(key=lambda x: (str(x["name"]).lower(), int(x["id"])))
    return rows


def main() -> None:
    m_path = DATA / "MTeams.csv"
    w_path = DATA / "WTeams.csv"
    if not m_path.is_file() or not w_path.is_file():
        raise FileNotFoundError(f"Need {m_path} and {w_path}")

    payload = {
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "sources": {"M": "MTeams.csv", "W": "WTeams.csv"},
        "M": _read_teams(m_path),
        "W": _read_teams(w_path),
    }

    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    print(f"Wrote {OUT} ({len(payload['M'])} men, {len(payload['W'])} women teams)")


if __name__ == "__main__":
    main()
